### PR TITLE
Refactor/schema move 2

### PR DESCRIPTION
**Pull request by creisle**
Updates the API to use the newly released schema v4

### CODE REVIEW

**Comment by mathieulemieux** [comment]
I tested the API with a new database instance and after having loaded some data with the loader.
When using the client (master branch, Node v12), the about page is showing this error:
image

**Comment by creisle** [comment]
> I tested the API with a new database instance and after having loaded some data with the loader. When using the client (master branch, Node v12), the about page is showing this error: image

Ya, that is a client error b/c the database is empty so there are no statements to show in the graph. Unrelated to the API

**Review by mathieulemieux — APPROVED** [review]
I did run the test suite, and I also used the client to do some queries, both while using the production-sync-dev and a new database. Everything went well. I did my best to follow changes made to files but I need more practice with the codebase to be comfortable enough with.

**Review by areisle — COMMENTED** [review]
I'm not super familiar with the api code, so my review is only be based on looking through the code changes and my knowledge of the schema package and will leave any manual testing to other reviewers

**Review by elewis2 — APPROVED** [review]

File: .eslintrc.json [hunk 1]
```diff
@@ -35,7 +35,16 @@
             "unix"
         ],
         "max-classes-per-file": "off",
-        "max-len": "warn",
+        "max-len": [
```
[the remark this answers is not in the document]
Comment (creisle) [reply .eslintrc.json:38]: ok, will do for future PRs

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@
 This repository is part of the [platform for oncogenomic reporting and interpretation](https://github.com/bcgsc/pori).
 
 The GraphKB database is implemented using [orientDB](https://github.com/orientechnologies/orientdb) and [orientjs](https://github.com/orientechnologies/orientjs).
-It is a graph database which is used to store variants, ontologies, and the relevance of this terms and variants. The KB uses strict controlled vocabulary to provide a parseable and machine-readable interface for other applications to build on. The API is a REST API built on node/express.
+It is a graph database which is used to store variants, ontologies, and the relevance of this terms and variants. The KB uses strict controlled vocabulary to provide a parsable and machine-readable interface for other applications to build on. The API is a REST API built on node/express.
```
[the remark this answers is not in the document]
Comment (creisle) [reply README.md:8]: ok, will do for future PRs

File: src/repo/commands/create.js [hunk 3]
```diff
@@ -49,23 +46,20 @@ const createUser = async (db, opt) => {
...
  *
  * @param {orientjs.Db} db the orientjs database connection
- * @param {Object} opt options
- * @param {Object} opt.content the contents of the new record
- * @param {string} opt.content.out the @rid of the source node
- * @param {string} opt.content.in the @rid of the target node
- * @param {ClassModel} opt.model the model for the table/class to insert the new record into
- * @param {Object} opt.user the user creating the new record
+ * @param {modelName} modelName the model for the table/class to insert the new record into
+ * @param {object} content the edge to be created
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/repo/commands/create.js:50]: The types will be removed from the comments when we move to TS (upcoming in the next 2 PRs) anyway

File: src/repo/schema.js [hunk 4]
```diff
@@ -103,24 +103,24 @@ const createSchema = async (db) => {
     await createSchemaHistory(db);
     // create the permissions class
     logger.log('info', 'create the Permissions class');
-    await ClassModel.create(schema.models.Permissions, db); // (name, extends, clusters, abstract)
+    await createModelInDb('Permissions', db); // (name, extends, clusters, abstract)
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/repo/schema.js:106]: I'm hoping we can use a type for that once we move to TS

File: src/repo/query_builder/fixed.js [hunk 5]
```diff
@@ -441,20 +447,20 @@ const keywordSearch = ({
...
     ...opt
 }) => {
-    const model = schema[target];
+    const model = schemaDefn.get(target);
 
     if (![OPERATORS.CONTAINSTEXT, OPERATORS.EQ].includes(operator)) {
-        throw new AttributeError(`Invalid operator (${operator}). Keyword search only accepts = or CONTAINSTEXT`);
+        throw new ValidationError(`Invalid operator (${operator}). Keyword search only accepts = or CONTAINSTEXT`);
     }
     if (!model) {
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/repo/query_builder/fixed.js]: fixed: bb7bf75

File: src/routes/index.js [hunk 6]
```diff
@@ -18,10 +18,10 @@ const parseClassListQueryParam = (param) => param.split(',').map((cls) => schema
 
 const addStatsRoute = (app) => {
     // add the stats route
-    const defaultClassList = Object.keys(schema).filter(
-        (name) => !schema[name].isAbstract
-            && schema[name].subclasses.length === 0 // terminal classes only
-            && !schema[name].embedded,
+    const defaultClassList = Object.keys(schemaDefn.models).filter(
+        (name) => !schemaDefn.models[name].isAbstract
+            && schemaDefn.descendants(name).length === 0 // terminal classes only
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/routes/index.js]: changed: ae7a3e891379f262d69eaf41445351ab38f59996

**Comment by mathieulemieux** [comment]
Tested with client(master) and loader(develop) on a new database. Tested with client(master&develop) on production-sync-dev. Everything seems to work.

**Comment by mathieulemieux** [comment]
Non-blocking: During review I saw that https://github.com/bcgsc/pori_graphkb_api/blob/c49d9e366ea5bf22749e8e700f53b5df7d087e0c/src/repo/commands/index.js#L8-L10 were no longer exported by ./src/repo/commands/select.js. Then they should probably not be imported and exported by ./src/repo/commands/index.js. Can we add these changes to the current PR ?

**Review by mathieulemieux — APPROVED** [review]

**Review by elewis2 — APPROVED** [review]

**Review by areisle — APPROVED** [review]

**Comment by creisle** [comment]
> Non-blocking: During review I saw that
> 
> https://github.com/bcgsc/pori_graphkb_api/blob/c49d9e366ea5bf22749e8e700f53b5df7d087e0c/src/repo/commands/index.js#L8-L10
> 
> were no longer exported by ./src/repo/commands/select.js. Then they should probably not be imported and exported by ./src/repo/commands/index.js. Can we add these changes to the current PR ?

In the interests of smaller PRs I will make a second PR for this

**Pull request merged**